File: part-3/perl/stream-stats.pl
#!/usr/bin/env perl

use strict;
use warnings;

my %stats;
while (<STDIN>) {
  chomp($_);
  my ($date, $time, $pid, $status, $msg) = split(/\|/, $_, 5);
  $msg =~ s/\|X*$//;
  $stats{$_}++ foreach ($msg =~ /\w+/g);
}

print "Word\tFrequency\n";
print($_, "\t", $stats{$_}, "\n") foreach ((sort {$stats{$b} <=> $stats{$a}} keys %stats)[0..9]);

File: part-3/python/stream-stats.py
#!/usr/bin/env python3

import sys
from functools import reduce
import pandas as pd

df = pd.read_csv(
  sys.stdin,
  sep="|",
  header=-1,
  names=["date", "time", "pid", "status", "data", "pad"]
)

words = []
for row in df["data"].str.split("[\W_]+"):
  for word in row:
    if len(word) > 0: words.append(word)

words = pd.DataFrame(words, columns=["Word"])
topwords = words.groupby("Word").size().nlargest(10)
print(topwords.to_string())
